### DISCUSSION TITLE
CSS-type styling for latex printing

### DISCUSSION THREAD

**Discussion by skieffer** [sympy/sympy#23589] [Ideas]
I don't know if there's an existing, quick and easy solution to #23574, but it reminded me of a project idea I wanted to discuss.

The `latex()` function already accepts a large nunber of kwargs, which give you various ways to control the form of the generated expression.

What if it accepted a new `style` kwarg, to which you could pass a dictionary, defining a whole set of styling rules, in a form analogous to CSS.

## Selectors

In CSS, you can select document elements by type (`a`, `p`, `div`, etc.), and also by "address". (In Firefox devtools, go to `Inspector`, right-click an element, and select `Copy > CSS Path`; in Chrome, use `Copy > Copy selector`.)

For mathematical expressions, there could be analogous ideas. We could use `Expr` subclasses like `Pow`, `Add`, etc. as "types,"
and expression elements already have natural "addresses" according to the tree structure of the expression.

## Example

For the `Pow` type, among the styling rules could be a `neg` key, whose legal values included `exp` ("put negative powers in the exponent") and `frac` ("write as a fraction, and move negative powers to the denominator").

Then for #23574, a solution would be:

```python
e = fps(5/(6-x))
latex(e, style={
    'Pow': {
        'neg': 'frac'
    }
})
```

Alternatively, the "address" approach could be used to make the styling more specific:

```python
latex(e, style={
    'function.1 Pow': {
        'neg': 'frac'
    }
})
```

Here, the overall expression `e` is a `Sum`, so `function` addresses the `Mul` whose factors
are `5`, `6^(-k-1)`, and `x^k`, and then `function.1` addresses the second factor `6^(-k-1)`.
So in this case the styling would be applied only to `Pow`s occurring within that one factor.
It makes no difference in this example, but it's not hard to imagine cases where it would matter.

**Comment by ritikBhandari** [comment]
I feel that's a good idea. In this way, we could have more authority over how we want our data to be displayed and under which constraints.

**Reply by ritikBhandari** [reply]
Though I think that could make things a little bit complicated for the first time. But still there are tradeoffs and I think in this case the pro outweighs con(not much of an issue).

**Reply by skieffer** [reply]
Whether this idea is overkill for the issue #23574 at hand is certainly a valid question. We could just give the `latex()` function a new kwarg like `neg_powers_denom` to control this particular formatting choice globally, throughout the expression.

But I think such one-by-one solutions are destined to multiply without bound, if we don't eventually introduce a comprehensive solution like the `style` kwarg.

**Reply by ritikBhandari** [reply]
exactly this is what I'm saying too. These types of scenarios can keep arising for different users who want their data to be displayed in a certain way. We can't keep building custom features every time a new issue arises. This one is really a comprehensive solution for future.

**Reply by ritikBhandari** [reply]
And also, I dont think so its really an 'overkill' per se. Rather it's an opportunity to efficiently manage such similar issues which certainly may pop up in the future.

**Reply by sylee957** [reply]
@ritikBhandari You should be aware of any technical complexity of introducing the options. 
There are already 10+ options and if we add via keyword argument, it gives a concern whether the options work in combination with other options. 
I just want to hear a proof about whether everything is compatible with everything

**Comment by asmeurer** [comment]
There's some prior art of using strings to select parts of an expression in the epathtools module, although I don't know if many people have used it.

**Reply by skieffer** [reply]
Great! So there's an existing language for this. I see it's [defined here](https://github.com/sympy/sympy/blob/ae598245cb778790d51b403fd9d6b6f850b4e438/sympy/simplify/epathtools.py#L10-L21) and there are some [examples here](https://github.com/sympy/sympy/blob/ae598245cb778790d51b403fd9d6b6f850b4e438/sympy/simplify/epathtools.py#L302-L309).

I suppose the `EPath` class would be useful for parsing rules defined by the user.

Meanwhile, the `LatexPrinter` class already achieves its own expression traversal. I imagine it would maintain a `self.epath` instance attribute, pushing and popping segments onto/off of this path as it entered and exited its various `_print_...()` methods.

**Reply by asmeurer** [reply]
To be sure, CSS is also a language for describing locations in trees. One difference with CSS is that it expects nodes to be labelled, which isn't the case in SymPy expressions. In any case, I think we should try to play around with these systems to see how well they work in practice. Like I said, I don't know if anyone actually uses epath. 

Also, the format doesn't have to be a string, it can be a Python object. Strings let us do terse things that wouldn't be valid Python, but Python objects allow specifying things that would be clumsy to express in a string (like a reference to specific object).

**Comment by oscarbenjamin** [comment]
I would like a whole different printing system where there are rewrite rules for all printing that users can override using pattern matching and replacement. The idea would be that there are symbolic representations of expressions and then separately symbolic representations of *displayed* expressions. The rewrite rules would govern how to convert from one to another e.g. `a*Pow(b, -1) -> DisplayDiv(a, b)`.

**Reply by asmeurer** [reply]
Somewhat related to this https://github.com/sympy/sympy/issues/20894.

**Comment by jaakkolehtinen** [comment]
I'd like to voice up in support of the general idea of being explicitly able to control how certain expressions are displayed.

Concretely, I'm currently playing with code that aims to explain probability computations in physically based rendering to my students, and I'm unpredictably getting negative exponents vs. fractions with positive exponents; only the latter makes the structure of the arising terms easy to see.

(Please note that I'm not yet familiar with Sympy deeply enough to know if there would be an easy fix for this particular case of mine; I also realize Sympy is likely not intended to be a pretty printer, so to speak.)

**Reply by skieffer** [reply]
I do have some old prototype code for this, which I've dusted off and made available [on PyPI](https://pypi.org/project/stylgebra/).

For example:

```python
>>> from sympy import symbols
>>> x = symbols('x')
>>> e = x**(-2)
>>> e
x**(-2)

>>> from stylgebra.adapt import adapt_sympy_expr
>>> f = adapt_sympy_expr(e)
>>> f.format(rules={'Power': {'negative': 'sup'}})
'x^{-2}'
>>> f.format(rules={'Power': {'negative': 'frac'}})
'\\frac{1}{x^{2}}'
>>> f.format(rules={'Power': {'negative': 'inline'}})
'1/x^{2}'
```

It has a long way to go, but it might already do enough for your use case.

**Reply by asmeurer** [reply]
You may also be interested in this issue https://github.com/sympy/sympy/issues/24834

**Comment by sylee957** [comment]
The problem is that the exact form of the expressions were never part of the API, 
so `Add, Mul` retains 3 arguments or `Pow` having in the second position isn't guaranteed across sympy versions
and you can easily recognize that there are avalanche changes in the test cases about the expressions forms,
when the intrinsic logic of solvers or simplify changes. 
So I'm afraid that the path based expression manipulation can stay too niche or can be fragile in that aspect.

I had often needed more delicate control of printing like that in practice, 
but I usually designed my own structures to contain sympy expressions, slot-by-slot, to control the printing as I want, 
than thinking of implementing that as option in sympy.

**Reply by sylee957** [reply]
My opinion is that we need to understand the characteristics of HTML elements than computer algebra objects.
Usually, HTML elements are not evaluated and its typesetting form stays same as the input, but usually computer algebra systems are not, and the form of the expressions does not always stay same.
